### PR TITLE
Make Korean IME input more consistent

### LINKED ISSUE #4226: Korean IME does not work as expected
# Environment

```none
Windows build number: Windows 10 2004 19041.21
Windows Terminal version (if applicable): 0.8.10091.0

Any other software?
```

# Steps to reproduce

- Run Windows PowerShell tab
- Typing text `$x = '한글'`

image

# Expected behavior

- The resulted text should be `$x = '한글'`

# Actual behavior

- But the resulted text was `$x = '한그ㄱ글'`

### PR DESCRIPTION
**Pull request by leonMSFT**
## Summary of the Pull Request
Korean IME was not working correctly due to way we were clearing the input buffer inside of `TSFInputControl`. We wanted to clear our input buffer and tell TSF to clear its input buffer as well when we receive a `CompositionCompleted` event. This works fine in some IME languages such as Chinese and Japanese. However, Korean IME composes characters differently in such a way where we can't tell TSF to clear their buffer during a `CompositionCompleted` event because it would clear the character that triggered the `CompositionCompleted` event in the first place.

The solution in this PR is to keep our `_inputBuffer` intact until the user presses <kbd>Enter</kbd> or <kbd>Esc</kbd>, in which case we clear our buffer and the TSF buffer. I've chosen these two keys because it seems to make sense to clear the buffer after text is sent to the terminal with <kbd>Enter</kbd>, and <kbd>Esc</kbd> usually means to cancel a current composition anyway.

This means we need to keep track of our last known "Composition Start Point", which is represented by `_activeTextStart`. Whenever we complete a composition, we'll send the portion of the input buffer between `_activeTextStart` and the end of the input buffer to the terminal. Then, we'll update `_activeTextStart` to be the end of the input buffer so that the next time we send text to the terminal, we'll only send the portion of our buffer that's "active".

## PR Checklist
* [x] Closes #4226
* [x] CLA signed
* [x] Tests added/passed

## Validation Steps Performed
Manual testing with Chinese, Japanese, and Korean IME.

### CODE REVIEW

File: src/cascadia/TerminalControl/TSFInputControl.cpp [hunk 1]
```diff
@@ -92,6 +93,27 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+    // Return Value:
+    // - <none>
+    void TSFInputControl::ClearBuffer()
+    {
+        if (!_inputBuffer.empty())
+        {
+            TextBlock().Text(L"");
+            _inputBuffer.clear();
+            _editContext.NotifyFocusLeave();
+            _editContext.NotifyTextChanged({ 0, ::base::ClampedNumeric<int32_t>(_inputBuffer.length()) }, 0, { 0, 0 });
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalControl/TSFInputControl.cpp]: we're clearing the input buffer before we get its length; won't it be 0 here?

**Comment by DHowett-MSFT** [comment]
@philnach if you've got a couple minutes, do you have any thoughts on this?

File: src/cascadia/TerminalControl/TSFInputControl.cpp (hunk 1 above)
Comment (leonMSFT) [reply src/cascadia/TerminalControl/TSFInputControl.cpp]: oof you're totally right 😞

Comment (miniksa) [reply src/cascadia/TerminalControl/TSFInputControl.cpp]: Oof, didn't see this one before I hit approve. Good catch, @DHowett-MSFT .

**Comment by DHowett-MSFT** [comment]
I've kicked just the x86 build to re-run; stand by.

**Comment by DHowett-MSFT** [comment]
I believe this made _emoji input_ do a weird thing: it only shows the most recently entered letter now

**Pull request merged**